addons/stock/models/product.py

```diff
@@ -761,6 +761,18 @@ def _onchange_type(self):
         return res
 
     def write(self, vals):
+        if 'company_id' in vals and vals['company_id']:
+            changing_company_products = self.filtered(lambda product: product.company_id.id != vals['company_id'])
+            if changing_company_products:
+                # Forbid changing the product's company when quants exist in another company.
+                existing_quants_count = self.env['stock.quant'].sudo().search([
+                    ('product_id', 'in', changing_company_products.product_variant_ids.ids),
+                    ('company_id', '!=', vals['company_id']),
+                    ('quantity', '!=', 0),
+                ], order=None, limit=1)
+                if existing_quants_count:
```

Comment:
I'm good with a single search! Let's just clean up the variable names:

```suggestion
            products_changing_company = self.filtered(lambda product: product.company_id.id != vals['company_id'])
            if products_changing_company:
                # Forbid changing a product's company when quant(s) exist in another company.
                quant = self.env['stock.quant'].sudo().search([
                    ('product_id', 'in', products_changing_company.product_variant_ids.ids),
                    ('company_id', '!=', vals['company_id']),
                    ('quantity', '!=', 0),
                ], order=None, limit=1)
                if quant:
```